src/mono/mono/metadata/class-init.c

```diff
@@ -605,31 +605,31 @@ mono_class_create_from_typedef (MonoImage *image, guint32 type_token, MonoError
 		switch (m_class_get_byval_arg (klass)->type) {
 			case MONO_TYPE_I1:
 				if (mono_defaults.byte_class)
-					klass->cast_class = mono_defaults.byte_class;
+					mono_defaults.byte_class->cast_class = klass;
 				break;
 			case MONO_TYPE_U1:
 				if (mono_defaults.sbyte_class)
-					mono_defaults.sbyte_class = klass;
+					klass->cast_class = mono_defaults.sbyte_class;
 				break;
 			case MONO_TYPE_I2:
 				if (mono_defaults.uint16_class)
-					mono_defaults.uint16_class = klass;
+					mono_defaults.uint16_class->cast_class = klass;
 				break;
 			case MONO_TYPE_U2:
 				if (mono_defaults.int16_class)
 					klass->cast_class = mono_defaults.int16_class;
 				break;
 			case MONO_TYPE_I4:
 				if (mono_defaults.uint32_class)
-					mono_defaults.uint32_class = klass;
```

Comment:
I don't understand why we were rewriting the mono_defaults  for `uint32_class` with the `int32` type.

Reply:
This code is inherit from legacy mono code.  It's wrong for years. but in practice mono run well because byte_class is initialized before sbyte, all mono_defaults.xxx_class will be assigned correctly finally so nobody found this issue.